File: packages/gavio-js/src/providers/openrouter.ts
/** openrouterAdapter — OpenAI-compatible Chat Completions API. */

import { ConfigurationError } from '../errors.js'
import type { PricingProvider } from '../pricing.js'
import type { GavioRequest } from '../request.js'
import type { GavioResponse } from '../response.js'
import { TokenUsage } from '../types.js'
import { BaseProviderAdapter } from './base.js'
import { postJson } from './http.js'

const DEFAULT_BASE_URL = 'https://openrouter.ai/api/v1'

export interface OpenRouterAdapterOptions {
  apiKey?: string
  baseUrl?: string
  timeoutMs?: number
  httpReferer?: string
  appTitle?: string
  pricing?: PricingProvider
}

export class OpenRouterAdapter extends BaseProviderAdapter {
  private readonly apiKey: string | undefined
  private readonly baseUrl: string
  private readonly timeoutSeconds: number
  private readonly httpReferer: string | undefined
  private readonly appTitle: string | undefined

  constructor(options: OpenRouterAdapterOptions = {}) {
    super(options.pricing)
    this.apiKey = options.apiKey ?? process.env['OPENROUTER_API_KEY']
    this.baseUrl = (
      options.baseUrl ??
      process.env['OPENROUTER_BASE_URL'] ??
      DEFAULT_BASE_URL
    ).replace(/\/+$/, '')
    this.timeoutSeconds = (options.timeoutMs ?? 30_000) / 1000
    this.httpReferer =
      options.httpReferer ??
      process.env['OPENROUTER_HTTP_REFERER'] ??
      process.env['OPENROUTER_REFERER']
    this.appTitle =
      options.appTitle ??
      process.env['OPENROUTER_APP_TITLE'] ??
      process.env['OPENROUTER_TITLE']
  }

  get providerName(): string {
    return 'openrouter'
  }

  url(): string {
    return `${this.baseUrl}/chat/completions`
  }

  headers(): Record<string, string> {
    if (!this.apiKey) {
      throw new ConfigurationError(
        'OPENROUTER_API_KEY not set (pass apiKey or set the env var)',
      )
    }
    const headers: Record<string, string> = { Authorization: `Bearer ${this.apiKey}` }
    if (this.httpReferer) headers['HTTP-Referer'] = this.httpReferer
    if (this.appTitle) headers['X-OpenRouter-Title'] = this.appTitle
    return headers
  }

  async complete(request: GavioRequest): Promise<GavioResponse> {
    const started = performance.now()
    const payload = {
      model: request.model,
      messages: request.messages,
      temperature: request.temperature,
      max_tokens: request.maxTokens,
    }
    const data = await postJson(
      this.url(),
      payload,
      this.headers(),
      this.timeoutSeconds,
    )
    const choices = (data['choices'] as Array<Record<string, unknown>>) ?? []
    const message = (choices[0]?.['message'] as Record<string, unknown>) ?? {}
    const content = (message['content'] as string) ?? ''
    const usageData = (data['usage'] as Record<string, number>) ?? {}
    const usage = new TokenUsage(
      usageData['prompt_tokens'] ?? 0,
      usageData['completion_tokens'] ?? 0,
    )
    return this.buildResponse(
      request,
      content,
      usage,
      (data['model'] as string) ?? request.model,
      started,
    )
  }

  async healthCheck(): Promise<boolean> {
    try {
      this.headers()
      return true
    } catch {
      return false
    }
  }
}

/** Factory: build an OpenRouter provider adapter. */
export function openrouterAdapter(
  options: OpenRouterAdapterOptions = {},
): OpenRouterAdapter {
  return new OpenRouterAdapter(options)
}

File: packages/gavio-py/gavio/providers/openrouter.py
"""OpenRouterAdapter — OpenAI-compatible Chat Completions API."""

from __future__ import annotations

import os
import time

from ..exceptions import ConfigurationError, ProviderUnavailableError
from ..pricing import PricingProvider
from ..request import GavioRequest
from ..response import GavioResponse
from ..types import TokenUsage
from ._http import post_json
from .base import ProviderAdapter

_DEFAULT_BASE_URL = "https://openrouter.ai/api/v1"


class OpenRouterAdapter(ProviderAdapter):
    """Talks to OpenRouter's OpenAI-compatible Chat Completions endpoint."""

    def __init__(
        self,
        api_key: str | None = None,
        base_url: str | None = None,
        timeout_seconds: float = 30.0,
        http_referer: str | None = None,
        app_title: str | None = None,
        pricing: PricingProvider | None = None,
    ) -> None:
        super().__init__(pricing)
        self._api_key = api_key or os.environ.get("OPENROUTER_API_KEY")
        self.base_url = (
            base_url or os.environ.get("OPENROUTER_BASE_URL") or _DEFAULT_BASE_URL
        ).rstrip("/")
        self.timeout_seconds = timeout_seconds
        self.http_referer = (
            http_referer
            or os.environ.get("OPENROUTER_HTTP_REFERER")
            or os.environ.get("OPENROUTER_REFERER")
        )
        self.app_title = (
            app_title
            or os.environ.get("OPENROUTER_APP_TITLE")
            or os.environ.get("OPENROUTER_TITLE")
        )

    @property
    def provider_name(self) -> str:
        return "openrouter"

    def url(self) -> str:
        return f"{self.base_url}/chat/completions"

    def headers(self) -> dict[str, str]:
        if not self._api_key:
            raise ConfigurationError(
                "OPENROUTER_API_KEY not set (pass api_key= or set the env var)"
            )
        headers = {"Authorization": f"Bearer {self._api_key}"}
        if self.http_referer:
            headers["HTTP-Referer"] = self.http_referer
        if self.app_title:
            headers["X-OpenRouter-Title"] = self.app_title
        return headers

    async def complete(self, request: GavioRequest) -> GavioResponse:
        started = time.monotonic()
        payload = {
            "model": request.model,
            "messages": request.messages,
            "temperature": request.temperature,
            "max_tokens": request.max_tokens,
        }
        data = await post_json(
            self.url(),
            payload,
            self.headers(),
            timeout=self.timeout_seconds,
        )
        choice = data["choices"][0]
        content = choice["message"]["content"] or ""
        usage_data = data.get("usage", {})
        usage = TokenUsage(
            prompt_tokens=usage_data.get("prompt_tokens", 0),
            completion_tokens=usage_data.get("completion_tokens", 0),
        )
        return self._build_response(
            request, content, usage, data.get("model", request.model), started
        )

    async def health_check(self) -> bool:
        try:
            self.headers()
            return True
        except ConfigurationError:
            return False
        except ProviderUnavailableError:
            return False
